### ISSUE TITLE
Leveraging NEP 18 for image processing and signal processing

### ISSUE THREAD

**Issue by jakirkham** [scipy/scipy#10204]
Am interested in how a user may leverage the image processing and signal processing APIs from SciPy, but implement the underlying algorithm to act on their own particular data structure without causing any changes in downstream user code. NEP 18 seems to suggest this is [one of the intended goals]( https://www.numpy.org/neps/nep-0018-array-function-protocol.html#use-outside-of-numpy ). Also [SciPy's roadmap alludes to this]( https://github.com/scipy/scipy/blob/v1.3.0/doc/source/roadmap.rst#support-for-distributed-arrays-and-gpu-arrays ) as well. Given it seems everyone is on-board with this idea, am really only curious how we get started? 😄

**Comment by jakirkham** [comment]
cc @shoyer @pentschev (who may also find this of interest and have thoughts here)

**Comment by rgommers** [comment]
Note that we have a GSoC starting for adding a backend system for `scipy.fft`. It will allow interfacing with Pyfftw, CuPy, Dask, mkl-fft, pypocketfft, etc. I'd like to see how that works out first, and see some real world experience with `__array_function__` for NumPy. So far my impression is that `__array_function__` is a useful band aid but not the right approach long-term.

**Comment by jakirkham** [comment]
Thanks for the info. That's great to hear. 😄

Working with FFT functions seems like a good starting place. FWIW Dask and CuPy support NEP 18 already and both will operate on their array types with their own FFT functions when NumPy's FFTs are used.

Would the idea be for SciPy to extend this to FFT functions it provides independently from NumPy (like discrete sine and cosine transforms)? Also do you have any thoughts on how real FFTs will be handled given the differences between NumPy and SciPy in this area? ( https://github.com/scipy/scipy/issues/2487 )

Interesting if `__array_function__` is not the right long-term approach, do you have any thoughts on what would be?

**Comment by pentschev** [comment]
My experience from the past few months with `__array_function__` is that it's useful in _most_ cases. However, there are some cases for which changes to existing code are necessary.

The most common case is creating new arrays with arbitrary shape, take for example `np.empty` (and its counterparts, `full`, `ones`, `zeros`). Because it doesn't take any arrays as reference for input, we have to create arrays differently. One such way is to use the new `empty_like` argument `shape` (see https://github.com/numpy/numpy/pull/13046), which takes an array as reference and allows us to override its shape.

I understand @rgommers point, and while I don't know if there are better ways we could do it long-term, `__array_function__` is not as straightforward as one would like, due to cumbersome cases such as the one I cited above.

**Comment by grlee77** [comment]
@jakirkham: See #10175 for some of the discussion relating to plans for the scipy FFT GSOC project. There has also been some discussion on the scipy-dev mailing list and in the text of @peterbell10's [proposal](https://docs.google.com/document/d/1hGMc6ZrsmSi_LSixI5oFbo5cZkPUo92sfmnVthg1u6Y/edit#).

One of the suggestions there (which I like) is to make the new FFT functionality available via `scipy.fft.*` rather than using the modifying the existing `scipy.fftpack` module. While we want to generally match the existing SciPy API, this would allow more flexibility to make some changes to things like `rfft` without breaking existing code.

**Comment by rgommers** [comment]
> while I don't know if there are better ways we could do it long-term, `__array_function__` is not as straightforward as one would like, due to cumbersome cases such as the one I cited above.

The protocol not covering things like array creation functions is one thing, but there's a more important limitation I think: it is specific to "types of arrays". So if you want to create functions with the same API for GPU arrays (CuPy, PyTorch), distributed arrays (Dask), sparse arrays (scipy.sparse, pydata/sparse), then it works. But what if you want to provide an alternative implementation for ndarrays? You simply cannot do that. Pyfftw, mkl-fft and pypocketfft all work on regular numpy arrays. So letting the numpy array carry around information about what implementation to use is just fundamentally not going to work. Instead, it's the library that must be able to say "hey, here's an implementation (perhaps for specific types)", and a mechanism for either automatic or user-controlled selection of which implementation/backend to use.

**Comment by jakirkham** [comment]
> ...there's a more important limitation I think: it is specific to "types of arrays".

Thanks for thinking about this issue, Ralf. Was about to ask what our solution should be for this case, but it sounds like you are one step ahead of me. 😉

**Comment by pentschev** [comment]
> Instead, it's the library that must be able to say "hey, here's an implementation (perhaps for specific types)", and a mechanism for either automatic or user-controlled selection of which implementation/backend to use.

Right, I had not considered such cases indeed. Do you already have ideas in mind on how to implement such mechanism? Perhaps it would be interesting to start a discussion (if it hasn't been started already).

As for array creation, you're right that it isn't indeed a limitation such as the types of arrays you mentioned, I was only pointing out that there's more work than just implementing `__array_function__`, and this cannot be neglected for large existing code bases, as I've recently discovered.

**Comment by rgommers** [comment]
@peterbell10's proposal (linked above) has some ideas. There's also http://uarray.readthedocs.io/, which is basically designed with exactly this purpose in mind.

**Comment by pentschev** [comment]
Thanks @rgommers, I'll check those out.

**Comment by rgommers** [comment]
> As for array creation, you're right that it isn't indeed a limitation such as the types of arrays you mentioned, I was only pointing out that there's more work than just implementing `__array_function__`, and this cannot be neglected for large existing code bases, as I've recently discovered

Yes, I agree. I have tried to use `__array_function__` (and `__array_ufunc__`) on some existing code bases (e.g. SciPy), and there's very few things that will work out of the box. For new code, I think you can carefully design it to work for numpy/dask/cupy.

**Comment by pentschev** [comment]
> For new code, I think you can carefully design it to work for numpy/dask/cupy.

Exactly what I was thinking right now. Perhaps we should document best practices for those `__array_function__` and `__array_ufunc__`. One piece that comes to mind right away is discouraging the use of `np.empty` (and counterparts) in favor of `np.empty_like(reference_array)`.

**Comment by peterbell10** [comment]
One big difference with what I've proposed is there would only be one active backend at a time. So, if you were using multiple different types of array, you would have to manually change the backend between calls. This avoids the overhead of a dispatch mechanism. It's also worth pointing out that, within this model, you could create a backend that itself does dispatching.

I'd be interested to hear any other opinions though.

**Comment by jakirkham** [comment]
I probably need to read your proposal more closely and give this more thought, but having one active backend at a time seems a bit tricky. At least in my mind there are two groups of arrays. First there are array data structure types (like NumPy, CuPy, Sparse, etc.). Second there are meta-arrays (like Dask, Xarray, etc.), which wrap arrays of the first kind. Adding a backend that only chooses one array type at a time might cause problems in getting both of these to interoperate. Do you have thoughts on how we might address this?

**Comment by jakirkham** [comment]
Small meta note: Sparse leverages NumPy internally to hold data related to its array type. So there may be a similar challenge to the one noted above in this specific case.

**Comment by jakirkham** [comment]
@peterbell10, are you going to be at SciPy this year? It might be good to find some time to discuss this issue if you are around.

**Comment by peterbell10** [comment]
>Adding a backend that only chooses one array type at a time might cause problems in getting both of these to interoperate.

Based on the dask issue referencing this, I take it you mean something like a meta-array of non-numpy arrays. That way the meta-array backend couldn't just defer to the default backend internally. That would indeed be tricky without at least some form of dynamic dispatch.

>@peterbell10, are you going to be at SciPy this year?

No, sorry.

**Comment by jakirkham** [comment]
Ah wouldn't worry about the xref. There was a subthread in that discussion that might benefit from your proposed solution.

That said, yes, Dask and Xarray are meta-arrays that contain various array types (NumPy, Sparse, CuPy, etc.) that may already rely on some kind of dispatching. In fact Xarray objects can be based off of Dask Arrays, which would add another meta level. Handling these meta-array levels would be useful.

No worries. Just think it would be useful to meet and hash this out at some point. Suppose we could do that virtually if that is an option for you. 🙂

**Comment by jakirkham** [comment]
@eteq, and I were talking about the convolve use case that would help bridge scikit-image, astropy, and dask-image.

**Comment by shoyer** [comment]
There are (at least) two different ways to use NEP 18 in SciPy:
1. Allowing SciPy functions to be overriden directly via `__array_function__`, e.g., `scipy.signal.convolve()` could directly check for `__array_function__` attributes.
2. Implementing SciPy functions in a duck-array compatible fashion via NumPy functions, e.g., `scipy.signal.convolve()` could avoid calling `np.asarray()`, and could be written in terms of `np.convolve` and `np.fft.fft` internally.

Which version are we talking about here? :)

**Comment by rgommers** [comment]
should be (1). (2) isn't the right thing to do (SciPy's `fft` is faster than NumPy's for example, so why rely on NumPy's) and not generically applicable anyway.

**Comment by jakirkham** [comment]
Yeah we are talking about 1. Other libraries may chose to handle the convolution differently. For example custom `NaN` handling came up in our conversation.

**Comment by jakirkham** [comment]
From discussions at SciPy with @rgommers, @danielballan, and myself, it sounds like we are going to wait on extending dispatching to other ndimage implementations until we've seen how the FFT use case plays out. Please feel free to add or correct things if I'm misremembering.

**Comment by yoav-palti** [comment]
Hi i am writing a DSP library (just starting) for personal usage and want that the library will be based on a Wave obj. Wave is suppose to be a ndarray subclass that also contains sample rate, main frequency and other things. I tried to use NEP 18 __array_function__ to override scipy convolve, correlate and fft. For the time being I saw they call asarray() witch makes the return type ndarray and I lose the metadata in the process. 
Is there any progress with this issue? Or any idea on how may I implement this?

**Comment by rgommers** [comment]
@yoav-palti removing `asarray` is going to take a long time, it's a big job. I do expect we'll make progress on this in 2021 though.

For now there isn't much you can do sanely. Although, if it's just for your own use, you could monkeypatch `np.asarray` to be a do-nothing op:
```
# write a content manager that temporarily replaces asarray to return its input rather than coerce to ndarray

with asarray_noop:
    signal.func(x_mywaveobj)
```

**Comment by yoav-palti** [comment]
I don't know the source coed of scipy very well but wont using asanyarray instead of asarray do the trick?

**Comment by rgommers** [comment]
That's not a good solution in general, because (a) that typically breaks with some subclasses (like `np.matrix`), and (b) it doesn't help with enabling other non-subclass array types like CuPy, Dask, etc. so it's not all that interesting to us.

**Comment by jakirkham** [comment]
Met with @jni @GenevieveBuckley @grlee77 @quasiben recently. We are interested in figuring out how we can manage dispatching here and how best we might make progress on this issue. @rgommers would you be able to advise on how we do this? Would it be helpful to setup a meeting to discuss? 🙂

**Comment by rgommers** [comment]
Hey @jakirkham, yes a meeting sounds good. I suggest to schedule one, and in parallel have a discussion on this issue about use cases. Scheduling a meeting is going to be fun between 3 continents - I can do until 11pm CEST, till midnight if really necessary. Will send you my availability now.

A few thoughts on this topic and where we are today with dispatching:

- The `scipy.ndimage` API could make sense to dispatch on, for two reasons: it's a sane and stable API, and it's all C code under the hood so making the functions themselves accept cupy/dask/etc arrays isn't really possible.
- Dispatching on the `scipy.signal` API I'm not a fan of, because it's a really badly designed API with lots of parts that need changing or deprecating. 

In terms of dispatching mechanisms:
- `__array_function__` status is more or less unchanged, it's useful when it works but it has limitations
- `uarray` has solved the meta-array dispatching problem, however its UX is clunky. That's solvable in principle, but the overall reception hasn't been enthusiastic.
- `__array_module__` (NEP 37) is the better dispatching principle long-term, and would allow (e.g.) JAX and PyTorch adoption. However not much progress has been made, and it's basically superceded by the array API standard (which effectively [adopts the same opt-in dispatching mechanism](https://data-apis.github.io/array-api/latest/purpose_and_scope.html#how-to-adopt-this-api) but adds a much better defined API).

All the above mechanisms apply to dispatching on an array API. Dispatching on, e.g., `scipy.ndimage` is a slightly different problem. The `__array_function__` approach you're asking for is: _"let me call `ndimage.some_func(x)` which let's `x.__array_function__` look up a compatible function, and if it finds one execute that instead of the SciPy function"_. There's no code reuse, instead there are two completely separate implementations accessed through the same API. 

The NEP 37 / array API standard approach is more: _"let's stop duplicating everything and reuse one implementation of a function by making it understand different array objects"_. This makes sense for pure Python + NumPy code; it may make sense for Cython code too (on CPU at least) but that's a problem to tackle later. 

The trouble with `__array_function__` is partly organizational. The right way to organize this if we designed it from scratch would be:
- a separate package `scipy-ndimage`
- that package supporting multiple array types
- all the dispatch to different "kernels" like a CPU one, a GPU one, and distributed CPU/GPU ones all within that one package

Instead, we now put the API in a single package, and each "kernel" in another package. This means the _implementation_ of a function effectively is spread across packages, which is very challenging when making changes to that function and really should only be considered for very mature and stable APIs.

I hope all the above makes sense. Can you describe in some more detail what functions you want to dispatch on? I assume it's scikit-image and cusignal related?

**Comment by GenevieveBuckley** [comment]
@rgommers we were talking primarily about dispatching scikit-image functions, so they could work with different input array types (dask, cupy, etc). But obviously this discussion is useful more broadly than that.

**Comment by rgommers** [comment]
Thanks @GenevieveBuckley.

CuPy has an `ndimage` API that's nicely compatible: https://docs.cupy.dev/en/stable/reference/ndimage.html. Dask not so much: http://image.dask.org/en/latest/dask_image.html. And JAX looks like it's aiming for full compatibility, but only has a single function so far (https://jax.readthedocs.io/en/latest/jax.scipy.html#module-jax.scipy.ndimage) and won't adopt `__array_function__` anyway. So it looks like either the `dask_image` API needs to change, or a more elaborate scipy-dask function-to-function mapping needs to live in `dask.array`. Or maybe the dispatching is for CuPy only, and you want to keep the Dask mapping in scikit-image?

**Comment by grlee77** [comment]
> Interesting I'd be curious to know what doesn't match in Dask-image

I think the functions themselves are largely compatible, but they are just organized in differently named submodules. 

The following table compares submodule names:

dask_image | scipy
-----|------
ndfilters | filters
ndmorph | morphology
ndinterp | interpolation
ndfourier | fourier

For example, SciPy users sometimes do `import scipy.ndimage as ndi` and then call functions from the submodules like (e.g. `ndi.morphology.binary_erosion`) or equivalently from the top level (e.g. `ndi.binary_erosion`), but neither of those styles currently work for `dask_image` since the submodule names are different and functions from the submodules aren't imported into the top-level `dask_image` namespace.

**Comment by jakirkham** [comment]
Interesting would have assumed module layout would be an internal detail. Is the module layout critical for dispatching or could that be relaxed? Are there any issues we foresee here?

**Comment by grlee77** [comment]
From a `scikit-image` standpoint, I see three types of functions:

1.) Those where NEP-18 (and possibly NEP-35) alone are already sufficient (either now or with minimal modification to scikit-image). These functions are in the minority for scikit-image. One trivial example is: [integral image](https://github.com/scikit-image/scikit-image/blob/main/skimage/transform/integral.py) and a less trivial one is [denoise_tv_chambolle](https://github.com/scikit-image/scikit-image/blob/59e4ddc8cc266b015ba907b221921b4865fa3243/skimage/restoration/_denoise.py#L396).

2.) Those where a combination of NumPy dispatching **and** `scipy.ndimage` dispatching would be sufficient. This is a substantially larger group of functions. Many functions in `skimage.filters`, `skimage.morphology` and `skimage.transforms` do not involve Cython or C code, but do rely internally on `scipy.ndimage` functions.

3.) The functions involving Cython/C/C++ code where dispatching to an entirely independent implementation is required. These probably require something like the `uarray` approach used for the `scipy.fft` module.

Use of SciPy modules other than `scipy.ndimage` or `scipy.fft` in scikit-image are less common. I think `scipy.signal.convolve`, `scipy.stats.entropy` and some of `scipy.sparse` are used a handful of places.

**Comment by grlee77** [comment]
> Interesting would have assumed module layout would be an internal detail. Is the module layout critical for dispatching or could that be relaxed? Are there any other issues we foresee here?

I was thinking of the case where SciPy itself does not do any dispatching, but third-party code tries dispatching based on importing the appropriate module depending on the input array type. e.g. like in the following example of a Gaussian smoothing function supporting both NumPy and CuPy arrays:
```Python
def dispatching_gaussian(image, sigma=5):
    if isinstance(image, numpy.ndarray):
        import scipy.ndimage as ndi
    elif isinstance(image, cupy.ndarray):
        import cupyx.scipy.ndimage as ndi
    return ndi.gaussian_filter(image, sigma=sigma)
```
There is no `dask_image` import that would give a module with the same API that can accept Dask arrays (where individual chunks could be either NumPy or CuPy arrays). One would have to import `gaussian_filter` directly instead of just using the module:

```Python
def dispatching_gaussian(image, sigma=5):
    if isinstance(image, numpy.ndarray):
        import scipy.ndimage as ndi
    elif isinstance(image, cupy.ndarray):
        import cupyx.scipy.ndimage as ndi
    elif isinstance(image, dask.array.core.Array):
        from dask_image.ndfilters import gaussian_filter
        return gaussian_filter(image, sigma=sigma)
    return ndi.gaussian_filter(image, sigma=sigma)
```

**Comment by jakirkham** [comment]
Yeah we could import these functions at a higher level or perhaps in an `ndimage` module so they are easy to get from one place

**Comment by GenevieveBuckley** [comment]
> So it looks like either the dask_image API needs to change, or a more elaborate scipy-dask function-to-function mapping needs to live in dask.array

I suspect changing the `dask-image` is the easiest and preferred way to go here. That's certainly do-able, if everyone agrees on the details.

It also sounds like if we get the API stuff sorted out with @rgommers to be the most compatible across the board, it might also clear up some [other discussions](https://github.com/dask/dask-image/issues/192#issuecomment-785550085) we are having.

**Comment by rgommers** [comment]
> For example, SciPy users sometimes do `import scipy.ndimage as ndi` and then call functions from the submodules like (e.g. `ndi.morphology.binary_erosion`) or equivalently from the top level (e.g. `ndi.binary_erosion`),

Please always prefer the latter - `ndimage` is the only public namespace, the rest only happens to miss an underscore in the filename.

> Interesting would have assumed module layout would be an internal detail. Is the module layout critical for dispatching or could that be relaxed?

For @grlee's case (2): the fully qualified name is used for lookup, see https://github.com/dask/dask/blob/67d8574a33265008ca69354011b44ea64c203a18/dask/array/core.py#L1503. If (sub)module name is different, then the fallback is an explicit dictionary mapping (`_HANDLED_FUNCTIONS`). There's no other way for dask's `__array_function__` implementation to know where to look.

> I was thinking of the case where SciPy itself does not do any dispatching, but third-party code tries dispatching based on importing the appropriate module depending on the input array type. e.g. like in the following example of a Gaussian smoothing function supporting both NumPy and CuPy arrays:

This is actually a lot less magical then adding `__array_function__` dispatching to `scipy.ndimage` directly. And it's easier for the API to evolve. For example, if we add a new keyword to an `ndimage` function, `dispatching_gaussian` won't accept it yet - you need to wait until `dask_image` supports it too, and then update `dispatching_gaussian` explicitly. Which is nicer than `ndimage.gaussian` accepting dask arrays, except for when that keyword is used (which will also be undocumented).

So it's worth considering that. Such a shim layer is easy to write. The advantage of `__array_function__` is only that new implementation of the API can be added - which makes some sense for array libraries (of which there are ~7 important ones, and then more array-like objects like unit-based arrays). But if all you have is CuPy, Dask and SciPy implementations for a single API (with no more in sight), then just handling those 3 explicitly will be more robust.

Note: this `dispatching_gaussian` type layer could easily live in `scipy.ndimage` itself. It just means that we'd explicitly become aware of CuPy and Dask arrays. Advantages over `__array_function__`-based dispatching are:
- `cupy` and `dask_image` can be optional dependencies, for which we can add CI. 
- much less backwards-incompatible to introduce
- can be documented explicitly
- numpy's `__array_function__` machinery isn't actually public, and has changes over supported releases (currently `>=1.16.5`). The explicit CuPy and Dask array handling will work with all numpy versions, and doesn't need the non-public machinery.
- `ndimage` API easier to evolve with new keywords
- in case `jax.scipy.ndimage` gains feature-completeness, it's extensible to JAX. JAX will not adopt `__array_function__`.

**Comment by rgommers** [comment]
Except this doesn't work:
```
elif isinstance(image, cupy.ndarray):
```
It should something like:
```
type(x).__module__ == 'cupy'
```
Needs some thought on what the most robust way of doing this. I don't know if for Dask there are array subclasses for example that one would want to pass such a check.

**Comment by jakirkham** [comment]
I think we are generally ok structuring things in dask-image so dispatching works.

Yeah `__array_function__` is just something we have in our toolbox today and is already used in a few libraries. It may not be strictly necessary for this use case.

The challenge with adding a new function is getting other libraries to buy-in to using this function. It's a little easier if it is covered through an existing function so things just work. For example what if we were to use something like [`singledispatch`]( https://docs.python.org/3/library/functools.html#functools.singledispatch )?

**Comment by GenevieveBuckley** [comment]
> Except this doesn't work:
> 
> ```
> elif isinstance(image, cupy.ndarray):
> ```
> 
> It should something like:
> 
> ```
> type(x).__module__ == 'cupy'
> ```

To be clear, this is no good because you want cupy to be an optional import (but you'd need to import it here to check with `isinstance`)? That's my understanding of why this is annoying/not possible, but if you meant something different, let me know.

> Needs some thought on what the most robust way of doing this. I don't know if for Dask there are array subclasses for example that one would want to pass such a check.

I don't think Dask arrays are commonly subclassed, If I'm wrong, can someone else jump in here with some examples?

The xarray DataArray class is not a subclass of dask array, but the `.data` attribute can contain a dask/numpy/etc. array.

The thing with Dask is that we often have two questions:
1. Is this a Dask array?
    ```python
    >>> type(a_dask_array).__module__
    'dask.array.core'
    ```
2. What type of chunks are in in this Dask array?
    We look at the type of the `_meta` attribute to find out this information.
    For example, a dask array with numpy chunks:
    ```python
    >>> type(a_dask_array._meta).__module__
    'numpy'
    ```

**Comment by rgommers** [comment]
> The challenge with adding a new function is getting other libraries to buy-in to using this function. 

With _"this `dispatching_gaussian` type layer could easily live in `scipy.ndimage` itself"_ I actually meant "we can make this the public API of `ndimage`. I.e. make the ndimage functions directly aware of CuPy and Dask. Which would not require changing any code in scikit-image or elsewhere.

> It's a little easier if it is covered through an existing function so things just work. For example what if we were to use something like [`singledispatch`](https://docs.python.org/3/library/functools.html#functools.singledispatch)?

`singledispatch` is basically a simpler version of `__array_function__` right? I think the tradeoffs are very similar.

**Comment by rgommers** [comment]
> To be clear, this is no good because you want cupy to be an optional import (but you'd need to import it here to check with `isinstance`)?

Exactly.

> The thing with Dask is that we often have two questions:

Thanks, that makes sense. SciPy would only need the first question.

**Comment by GenevieveBuckley** [comment]
@rgommers what do you need to have happen within dask-image to support this effort?

Making sure that all of the dask-image functions are imported into the top level namespace? Anything else?

> There is no dask_image import that would give a module with the same API that can accept Dask arrays (where individual chunks could be either NumPy or CuPy arrays). One would have to import gaussian_filter directly instead of just using the module

@jakirkham It sounds like you have a structure in mind, but I'm not completely sure what it is. Should we:
1. Import all functions into the top level namespace so we can do ` import dask_image as ndi`
2. Create a new `ndimage` module that does nothing except make all the ndimage functions available through `import dask_image.ndimage as ndi` (and still allowing access as usual for backwards compatibility, eg: `from dask_image.ndfilters import gaussian_filter')
3. Put all of the functions into a new `ndimage` subpackage (breaking change)

**Comment by rgommers** [comment]
Not necessarily the top-level namespace, but a namespace that matches. If you'd want to name it ``dask_image.scipy` or `dask_image.scipy.ndimage`, that should work too.

**Comment by jakirkham** [comment]
Sorry for the delay. Have been pretty swamped over the past couple weeks 😅

> > The challenge with adding a new function is getting other libraries to buy-in to using this function.
> 
> With _"this `dispatching_gaussian` type layer could easily live in `scipy.ndimage` itself"_ I actually meant "we can make this the public API of `ndimage`. I.e. make the ndimage functions directly aware of CuPy and Dask. Which would not require changing any code in scikit-image or elsewhere.

I think this is the most important thing honestly. Having this as an API that can be used with different array libraries and doing the right thing under-the-hood is the most important thing

> > It's a little easier if it is covered through an existing function so things just work. For example what if we were to use something like [`singledispatch`](https://docs.python.org/3/library/functools.html#functools.singledispatch)?
> 
> `singledispatch` is basically a simpler version of `__array_function__` right? I think the tradeoffs are very similar.

I'm not sure that I would compare it to `__array_function__` exactly. Every form of dispatching we are looking at here looks somewhat similar to others.

With `singledispatch` downstream libraries would need to explicitly register their callback with their array type. This handles the `isinstance` check, but saves SciPy from thinking about this. We could also have a generic fallback for `object` that would error or do something else reasonable. It also relaxes the need to match a particular structure and allows libraries to figure out how much or little glue code they want to implement this. Here's one way it might be used, but not the only way

```python
from functools import singledispatch

@singledispatch
def smooth(arr):
    typ = type(arr)
    raise NotImplementedError(f"Support for `{typ}` is not implemented")

@smooth.register(numpy.ndarray)
def _(arr):
    # do something with a NumPy array
    ...

@smooth.register(cupy.ndarray)
def _(arr):
    # do something with a CuPy array
    ...
```

**Comment by rgommers** [comment]
`ndimage` has several functions that take multiple array-like inputs, so `singledispatch` seems suboptimal. I had a look at the history of dispatch frameworks, and it was something like `singledispatch` -> `multipledispatch` -> `__array_function__` / `uarray`. This needs a little more thought about the trade-offs. Being able to evolve the API (adding new keywords, adding new functions) without things breaking or being silently unsupported seems to be most important.

For the next week I won't have time for this, I'll try to revisit the week after.

**Comment by grlee** [comment]
Can you remove me from this thread?

George Lee

**Comment by jakirkham** [comment]
Sorry George. I think someone probably mistyped Greg's GH handle earlier. Unfortunately we can't remove you ourselves, but you can click the "Unsubscribe" button above to the right to remove yourself. We will try to be more careful in the future. Thanks for letting us know and sorry for the noise

<img width="1792" alt="Screen Shot 2021-03-24 at 5 45 29 PM" src="https://user-images.githubusercontent.com/3019665/112401908-bc928980-8cc8-11eb-85d5-4a9abb497b87.png">

**Comment by rgommers** [comment]
I edited my offending comment with the wrong user handle. Will remove the comments about this topic later.

**Comment by rgommers** [comment]
I've been thinking about this some more today. `uarray` really does seem like the most appropriate solution here (xref [last years' meeting minutes with a comparison](https://github.com/numpy/archive/blob/e444cab6e26501363c2187d13cfbd4a6397cee83/other_meetings/2020-04-21-array-protocols_discussion_and_notes.md#comparisons-table-initially-hameers)). The two main questions I have right now that I don't really know/remember the answer to are:

1. What happens if a function which supports uarray gains a new keyword, and then a user calls it with (e.g.) a cupy array when `cupyx.scipy.ndimage` does not yet have that keyword?
2. Can we change `uarray` to let backends auto-register, so we don't have to make end users use context managers? IIRC we only avoided auto-registration because NumPy devs didn't like it and there was an attempt to have NumPy adopt `unumpy`.

(1) is the one thing of my list of desirable traits in https://github.com/scipy/scipy/issues/10204#issuecomment-788764360 that we'd get from explicit forwarding. (2) is the main UX problem with `uarray`, and it's bad enough that it ruins the experience of using it despite all the nice features and performance it offers.

**Comment by rgommers** [comment]
@hameerabbasi or @peterbell10 any thoughts on these uarray questions?

**Comment by hameerabbasi** [comment]
Anyone can Auto register today, it's only a question of changing it from "discouraged" to "allowed practice".

I'd be on board with auto registration, but the order in which backends will be called will be undefined.

**Comment by rgommers** [comment]
> I'd be on board with auto registration, but the order in which backends will be called will be undefined.

That's not a real problem other than a bit of extra overhead, right? We don't need to support mixed-array-kind inputs in a single function call.

**Comment by jakirkham** [comment]
@m-albert, pointed out that some ndimage functions can be written in terms of others ( https://github.com/dask/dask-image/issues/201#issuecomment-822049649 ). So maybe not all functions need to dispatch as long as they can be rewritten in terms of others that do dispatch

**Comment by hameerabbasi** [comment]
> > I'd be on board with auto registration, but the order in which backends will be called will be undefined.
> 
> 
> 
> That's not a real problem other than a bit of extra overhead, right? We don't need to support mixed-array-kind inputs in a single function call.

Yes, if backends do the right thing, the behaviour remains unchanged.

**Comment by rgommers** [comment]
> @m-albert, pointed out that some ndimage functions can be written in terms of others ( [dask/dask-image#201 (comment)](https://github.com/dask/dask-image/issues/201#issuecomment-822049649) ). So maybe not all functions need to dispatch as long as they can be rewritten in terms of others that do dispatch

While this is possible, I'm really not a fan of such things. It's fragile because it couples function dispatch with implementation details in the function body. A clean separation for every function between dispatching first, and continuing within the library if no dispatchable object is encountered second is better. 

It isn't that much work to write the dispatching code, so we should just do it for every function.

**Comment by hameerabbasi** [comment]
> While this is possible, I'm really not a fan of such things. It's fragile because it couples function dispatch with implementation details in the function body. A clean separation for every function between dispatching first, and continuing within the library if no dispatchable object is encountered second is better. 

I'd like to note here that `uarray` already has functionality for this: It's called a "default implementation". One can provide such an implementation and it's treated as a fallback.

**Comment by j-bowhay** [comment]
Closing as we now have array API support
